### PR TITLE
Attempt to articulate rationale for closing stale issues

### PR DESCRIPTION
**Pull request by bkeepers**
This has been raised enough times that I think it's worth documenting.

I wanted to hit on these points:

- there are finite resources
- a mountain of issues is paralyzing
- this makes the implicit explicit
- it only takes a comment to keep the conversation going

cc people who I know I've talked with this about: @JasonEtco @gr2m @lee-dohm @mikemcquaid

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -82,6 +82,14 @@ An easy way to check and see which issues or pull requests will initially be mar
 
 To avoid triggering abuse prevention mechanisms on GitHub, only 30 issues and pull requests will be marked or closed per hour. If your repository has more than that, it will just take a few hours or days to mark them all.
 
+## Is closing stale issues really a good idea?
+
+In an ideal world with infinite resources, there would be no need for this app.
```
Comment (JasonEtco) [inline README.md]: I don't think that this is necessary - it feels a little defensive (which I understand).

File: README.md [hunk 2]
```diff
@@ -82,6 +82,14 @@ An easy way to check and see which issues or pull requests will initially be mar
...
+
+But in successful software project, there's always more work to do than people to do it. As more and more work piles up, it becomes paralyzing. Just making decisions about what work should and shouldn't get done can exhaust all available resources. In the experience of the maintainers of this app—and the hundreds of other projects and organizations that use it—focusing on issues that are actively affecting humans is an effecitive method for prioritizing work.
+
+To some, a robot trying to close stale issues may seem inhospitable or offensive to contributors. But the alternative is to disrespect them by setting false expectations and implicitly ignoring their work. This app makes it explicit: if work is not progressing, then it's stale. A comment is all it takes to keep the converstion alive.
```
Comment (JasonEtco) [inline README.md]: Typo in last sentence, `converstion`

File: README.md [hunk 3]
```diff
@@ -82,6 +82,14 @@ An easy way to check and see which issues or pull requests will initially be mar
 
 To avoid triggering abuse prevention mechanisms on GitHub, only 30 issues and pull requests will be marked or closed per hour. If your repository has more than that, it will just take a few hours or days to mark them all.
 
+## Is closing stale issues really a good idea?
+
+In an ideal world with infinite resources, there would be no need for this app.
+
+But in successful software project, there's always more work to do than people to do it. As more and more work piles up, it becomes paralyzing. Just making decisions about what work should and shouldn't get done can exhaust all available resources. In the experience of the maintainers of this app—and the hundreds of other projects and organizations that use it—focusing on issues that are actively affecting humans is an effecitive method for prioritizing work.
```
Comment (JasonEtco) [inline README.md]: Typo in first sentence - `project` should be `projects`, or in `_a_ successful software project`.

**Comment by lee-dohm** [comment]
I attempted to write up an addition to Atom's CONTRIBUTING guide with this kind of information a couple days ago. Here's what I came up with (but haven't finished polishing) ...

----

We use Probot stale to close issues that haven't seen _any_ activity in over a year. This can happen for many reasons. Most often it is one of the following:

* We fixed the problem and forgot to close the issue
* We fixed the problem and the person fixing it didn't know there was an associated issue
* The problem went away (because it was a temporary problem caused by something outside of Atom's control) and we forgot to close it
* The issue author fixed or found the real source of the problem and forgot to come back to close it
* We asked the issue author for more information, they never got back to us and we forgot to come back to close it

You'll notice that "we forgot to close it" is a part of many of the above reasons. It would be great if we had all the time we needed to keep our issues list tidy and we remembered to close everything that needed it. Unfortunately, the Atom maintainer team consists solely of humans (so far) who have a lot of things that we have to and want to do. Because we value our community, we naturally prioritize responding to new issues over going through old issues. So we have Probot stale help clean up after us.

**Comment by patcon** [comment]
This is great! So happy this conversations is happening :)

I posted a thing here, that might give some context on the perspective I bring: https://github.com/probot/settings/issues/41#issuecomment-378813957

General thesis of rant: Stalebot -- in dutiful service of _current_ human resource constraints -- feels like it unthinkingly prioritizes discard of _past_ contributors' time resources, over the mobilization of new human resources.

The one thing I retract is that I don't "dislike the thing made", but rather, I feel like it defaults to some practices that can too easily encourage community anti-patterns. Or perhaps, it's that it doesn't suggest itself as an intermediary solution in a larger process -- like a lichen or moss, or another pioneering species in an [ecological succession](https://en.wikipedia.org/wiki/Ecological_succession) -- making space for more nuanced and diverse species and strategies and niches in the future :)

Curious if other vibe on the above, of take issue with it.

Anyhow, I'll come back later. I promise I have real suggestions and not just "get me a taller soapbox!" :slightly_smiling_face:

File: README.md (hunk 1 above)
Comment (patcon) [reply README.md]: fwiw this statement really connects with me -- i saw @bkeepers mention it elsewhere, and it humanizes the effort

**Comment by lee-dohm** [comment]
> prioritizes discard

Nothing is discarded. Nothing is deleted. Everything stays right where it is. Closing an issue in most mature projects means that it is not currently under consideration for taking action by the maintainer team. It could be reopened at any moment given the right information, the right stimulus, or the right justification or perspective. The idea that closing an issue is a destructive action is a fallacy that should be discouraged.

**Comment by patcon** [comment]
Thanks for engaging, @lee-dohm. I'd like to respectfully disagree with the **premise of "destruction"** being the right one. (My bad if I implied that framing at any point btw -- it doesn't seem right.

 By way of analogy: Is putting something in the rafters of your garage a destructive action? Where we choose or encourage communities to keep things matters -- feature requests or careful thoughts about our projects -- or yes, garbage and lazy bug reports. Is it on the fridge, on the workbench, or in the rafters of the garage? (There is no garbage can on github, even for the true garbage.)

I don't have the answers, but I think it's a more nuanced convo than you're giving it credit for

**Comment by lee-dohm** [comment]
To extend your analogy a little further, putting something in the rafters of your garage is not "discarding" it. Like a parent packing away old athletic trophies for safe keeping in case you decide to one day reminisce, it is respectfully holding on to those memories and achievements, just not necessarily keeping them on the mantlepiece anymore. Perhaps you feel slighted because they are not on the mantlepiece, but the mantlepiece is only so big and keeping too many trophies there devalues all of them risking making all of them meaningless. Whereas keeping the most significant or recent ones on the mantlepiece and putting the rest in the rafters attributes the appropriate value to each one. And to risk stretching this analogy to the breaking point, if some new piece of information comes to light about or some new significance is ascribed to one of those trophies in the rafters, it can easily be retrieved and placed back on the mantlepiece at any time.

**Comment by patcon** [comment]
You used a really emotional examples, but the bot can't know/judge that context. I would love to see it be less agressive by default, and act more like it should with the very little that it knows.

A maintainer of a 400-open-issue project can and maybe should have stronger opinions, but I wish the bot would not impose that on all users, big and small. Defaults are among the most political aspects of technology. <3

Back to analogy, I hope for a bot that does not default to auto-moving things from fridge to living room to garage on a set timeline, as if all homes are factories and have those requirements :)

**Comment by patcon** [comment]
Heh, I feel like this conversation needs a Loomio thread, and not a GitHub issue/PR. I apologize.

Does GitHub staff ever use Loomio?

**Comment by patcon** [comment]
PS - I don't mean to imply that using stalebot is a open-source community sin, and I'm not advocating for removal of any features.

**Comment by lee-dohm** [comment]
> You used pretty emotional examples, which is ok, but I'd like to point out that the bot can't know/judge that context.

I would argue that I extended the already emotional analogy that you created by likening closing an issue to placing it in the rafters in the garage. I certainly reframed the emotional context to support my point-of-view, but that's the subtle technique behind using analogies, isn't it? :grinning:

The point behind my extension of the analogy is that I suspect the typical open source contributor would likely experience the same emotional reaction to the same action whether a bot carried out the action or a human carried it out. One might feel that the action is a bit _more_ impersonal because a human "couldn't be bothered to tell me to my face and used a bot to do it instead", but I feel that's a question of degree not of kind.

The underlying source of the emotional reaction is that people feel their contribution is being "thrown away" or "discarded", which ultimately means they feel their contribution is not being ascribed the value they think it deserves. This is a quite common reaction in open source projects and one that maintainers have to deal with on a nearly constant basis. It is also one that, frankly, causes a lot of stress for a lot of maintainers. This stress can ultimately lead to maintainer burnout if they can't manage to create a sense of emotional detachment and realize that the contributor's emotional well-being isn't the maintainer's responsibility so long as the maintainer is being polite, open, and honest. Having a bot deliver a message on behalf of a maintainer is perhaps one way to create that emotional distance.

This is all really complex stuff that we haven't figured out best practices for. But I feel that Probot stale is:

1. A step in the right direction
1. Allowing us to explore all the different nuances that will help us suss out the best practices at a scale that we couldn't achieve before
1. Allowing us to explore those nuances in an environment of greater safety for all involved, contributors and maintainers

**Comment by patcon** [comment]
heh I'll concede that analogies often have a baseline emotional manipulation element :slightly_smiling_face:

(If this message says nothing new, please do tell me that. I'm fully engaged here and trying, but if I sound like a broken record, then I'm happy to vote myself off the island.)

Real quick: I'm also trying really hard to be introspective and self-aware, and so I think I can say this honestly: This is emphatically _not_ about my one issue :) I will acknowledge that it wasn't even a great synthesis!

A prime concern for me is that `probot/stale` has these **harsh opinionated character traits by default** --
ie doesn't just mark, but marks as `wontfix`, then eventually closes -- whether activated on something like nodejs core, or on new software dev Sally's first 10-star project. This plugins could be about reactivating the humans, but instead it feels from the language and defaults that it's about marking the garbage and clearing it out. That feels not-so-great imho

> likely experience the same emotional reaction to the same action whether a bot carried out the action or a human carried it out. 

Respectfully disagree. This does not connect with my experience in the world. Most people don't like to be managed by technology. I feel that `probot` technology will be at its best when it's more akin to a social exoskeleton (for maintainers and the wider community), and less like a patrolbot guarding the maintainers.

> not being ascribed the value they think it deserves

This next comment from me feels a little too self-satisfying to write, and so I have a little skepticism of its merit, but: Asking for _any_ value does not seem to be asking very much. closed issues are the garbage pile of our ecosystem -- it's as far away as we can send something in userspace. I can only speak for how I explore new projects, but I don't think it's in our cultural practices to pick through that pile.

> Having a bot deliver a message on behalf of a maintainer is perhaps one way to create that emotional distance.

I hear you on this.

---

This is really about the philosophy. I know `probot/stale` must exist now to solve an immediate and important need: the mental well-being of maintainers of large projects.

But it's not the ideal solution. The ideal solution is in automation that helps humans to create welcoming spaces for contributors to step forward, feel welcome, and stick around. Great open source automation calls in reinforcements. `probot/stale` sends people away and puts their contributions as far away as we can put them, all without inviting more humans to help, and cuts a final tie between prospective "leads" who've already shown up (the issue filer and participants), in a way that feels like a waste. And it's not even that the issue creator is the only human capital opportunity lost. The people who could help triage are also the human capital, and the bot is automating away the easiest sort of human'ing -- the low-hanging fruit of community building; the first rung in the contributor engagement ladder -- in a strategy that seems to cut the legs out from under itself.

Strongly agree on your final 3 points :heart: 

Agreeing on "step in right direction" might seem odd -- I expressed earlier that I didn't believe `probot/stale` was a winning path. But I think this convo has **helped me re-evaluate to see `probot/stale` more like a part of that ecological succession** I mentioned -- lichens to grass to shrubs to oaks. Open source communities need to be healthier, and I hope tactics like `probot/stale` can be a band-aid to give us mental space to collect ourselves and find the better things -- more humane automation.

:heart: :heart: :heart:

**Comment by lee-dohm** [comment]
> if I sound like a broken record

No, I don't feel like you sound like a broken record. We're both refining our points to zero in on the underlying motivations. That's what good discussions do :grinning:

> Anyhow, I think I'm going to cut myself off 🤐

Fair enough.

> I will admit to feeling a little bummed to find myself so misaligned with maintainers of projects that I love and respect, with a sense of helplessness around how to bridge understanding

Just to let you know, I've shared this discussion around to a number of my contacts that are interested in open source as an illustration of the dichotomy of motivations between the maintainer of and the contributor to open source projects. So perhaps you can take :heart: in that? It is natural for smart, capable people coming at the same problem from different directions to see things in different ways and finding the correct middle ground is always challenging.

> hanks for the generosity of your time if this interaction was stimulating

It most definitely was. Thank you :bow:

**Comment by MikeMcQuaid** [comment]
> Respectfully disagree. This does not connect with my experience in the world. Most people don't like to be managed by technology. I feel that probot technology will be at its best when it's more akin to a social exoskeleton (for maintainers and the wider community), and less like a patrolbot guarding the maintainers.

To chime in with Homebrew's example (one of the most active projects on GitHub with large numbers of contributors and a project I've maintained for almost 9 years): people massively prefer stalebot closing their issues to having a maintainer do it with the same category. People respond better to positivity and empathy that come from humans rather than bots and people respond better to pedantry and rules to be enforced by bots rather than humans (because they are seen as less arbitrary). People may not _like_ to have bots close their issues but they will react better than a specific maintainer doing so.

As a wider comment: issues that are not getting fixed and PRs that are not getting merged should just be closed. There's no point in leaving them open forever. Where there's projects where good, green CI PRs with all comments addressed are sitting and not getting merged I agree probot-stale may reveal that that project has other issues.

**Comment by patcon** [comment]
Thanks @MikeMcQuaid! :)

I realize that what I'm attempting to speak to is largely about **defaults**. For that conversation on the right choices, I am wary that it might only be attended by self-selecting large maintainers who feel pain -- which is a little akin to how larger businesses feel the pain of regulation more, and make their voices heard. (I regret that this analogizes maintainers to companies, as I'm not trying to say anything about that here!) With lobbied changes to regulation based on the needs of big businesses, the **long-tail of private citizens will never feel the loss to "the global commons" in a way that's strong for them to show up**.

Similarly, small and starter maintainers won't show up here. They don't feel it strongly enough. They just receive the tools that we give them, and take that to be the world. But those tools (and their defaults) shape character and community :)

So I'm trying to speak for how some defaults (tailor-made to large maintainers' needs) will not serve small projects, and may in fact hurt them in how they grow.

There' s a term in community organizing that goes something like, **"You shouldn't relay the problems of the butterfly to the world of the caterpillar"**. I would love for the stalebot maintainer community (and maybe the wider audience building the tech glue for our communities) to be aware of this dynamic in offering up this platter of tools -- heck, even in offering all tools we create in the tech community <3

**Comment by bkeepers** [comment]
Thanks everyone for the lively discussion. I'm going to merge this as-is, but would appreciate PRs to add nuance, caveats, or points that were not included.

**Pull request merged**